### PR TITLE
Fix User-Agent usage in SignalR

### LINKED ISSUE #29183: grpc-dotnet + Blazor sets Safari and Firefox User-Agent to "grpc-dotnet/$assemblyVersion"
### Describe the bug

When using grpc-dotnet and Blazor together to call a GRPC API, the User-Agent is set to "grpc-dotnet/$assemblyVersion" in Safari and Firefox. In Chrome and Edge it is the actual browser User-Agent.

https://docs.microsoft.com/en-us/aspnet/core/grpc/services?view=aspnetcore-5.0 for example describes how to read the User-Agent:

```cs
public override Task<ExampleResponse> UnaryCall(ExampleRequest request, ServerCallContext context)
{
    var userAgent = context.RequestHeaders.GetValue("user-agent");
    // ...

    return Task.FromResult(new ExampleResponse());
}
```

(Migrated from https://github.com/dotnet/AspNetCore.Docs/issues/21178)

### To Reproduce

Use GRPC with Blazor as described at https://docs.microsoft.com/en-us/aspnet/core/grpc/browser?view=aspnetcore-5.0 and look at the traffic in Safari. The User-Agent will be for example "grpc-dotnet/2.34.0.0".

You can also check out https://our.gatekeeper.page in Safari or Firefox and look at the request to the "WhoAmI" endpoint to see the header being wrong. The code for this can be found at https://github.com/getgatekeeper/server.

The Blazor example from https://github.com/grpc/grpc-dotnet/tree/master/examples/Blazor should also have the same issue.

The issue seems potentially related to the following piece of code in [`Grpc.Net.Client/Internal/GrpcCall.cs`](https://github.com/grpc/grpc-dotnet/blob/13a4bf1b2dea216c055370a8ec89827e0635f75d/src/Grpc.Net.Client/Internal/GrpcCall.cs#L798-L799). I wonder if Chrome/Edge just ignore the header being overwritten whilst Safari and Firefox obeys it: 

```cs
// User agent is optional but recommended.
headers.TryAddWithoutValidation(GrpcProtocolConstants.UserAgentHeader, GrpcProtocolConstants.UserAgentHeaderValue);
```

## Test plan
| Browser  | Sends correct User-Agent | Screenshot |
| ------------- | -------------  | ------------- |
| Safari  | ❌  |image| 
| Firefox  | ❌  |image| 
| Chrome  | ✅   |image|
| Edge  | ✅   |image|

### LINKED ISSUE #30137: [SignalR] Could not parse value for header 'User-Agent'
### Describe the bug
When i try to connect from a `Microsoft.AspNetCore.SignalR.Client` with a SignalR server i get the following exception:
`Failed connecting with the SignalR server: Could not parse value for header 'User-Agent'`
What i have found so far is that:
- This does not happen when i set the transfer protocol to `WebSockets` **AND** `SkipNegotiation` is true ex:
```CSharp
_client = new HubConnectionBuilder()
    .WithUrl($"http://{serverIp}:{port}/client", HttpTransportType.WebSockets, options =>
    {
        options.SkipNegotiation = true;
    })
    .Build();
```
- It seems to be related to Mono. I am trying to make this work with Godot game engine which depends on mono.
- It seems to be related to how the Mono HttpClient serialises the headers?
### To Reproduce
Try to connect with a server using mono runtime.

### Exceptions (if any)
```CSharp
at bool System.Net.Http.Headers.HttpHeaders.AddInternal(string name, IEnumerable<string> values, HeaderInfo headerInfo, bool ignoreInvalid)
at void System.Net.Http.Headers.HttpHeaders.Add(string name, IEnumerable<string> values) x 2
at HttpClient Microsoft.AspNetCore.Http.Connections.Client.HttpConnection.CreateHttpClient()
at new Microsoft.AspNetCore.Http.Connections.Client.HttpConnection(HttpConnectionOptions httpConnectionOptions, ILoggerFactory loggerFactory)
at async ValueTask<ConnectionContext> Microsoft.AspNetCore.Http.Connections.Client.HttpConnectionFactory.ConnectAsync(EndPoint endPoint, CancellationToken cancellationToken)
at async Task Microsoft.AspNetCore.SignalR.Client.HubConnection.StartAsyncCore(CancellationToken cancellationToken)
at async Task Microsoft.AspNetCore.SignalR.Client.HubConnection.StartAsyncInner(CancellationToken cancellationToken)
at void System.Threading.Tasks.ForceAsyncAwaiter.GetResult()
at async Task Microsoft.AspNetCore.SignalR.Client.HubConnection.StartAsync(CancellationToken cancellationToken)
at async void IO.Client.Singletons.Connection.ServerConnection.ConnectWithServer() in XXXX
```

### Further technical details
```
.NET SDK (reflecting any global.json):
 Version:   5.0.200-preview.20614.14
 Commit:    863605c8c3

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.200-preview.20614.14\

Host (useful for support):
  Version: 5.0.2
  Commit:  cb5f173b96

.NET SDKs installed:
  2.1.517 [C:\Program Files\dotnet\sdk]
  3.1.402 [C:\Program Files\dotnet\sdk]
  5.0.100-preview.8.20417.9 [C:\Program Files\dotnet\sdk]
  5.0.102 [C:\Program Files\dotnet\sdk]
  5.0.200-preview.20614.14 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.21 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.21 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0-preview.8.20414.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.21 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0-preview.8.20407.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.8 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0-preview.8.20411.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```
My question is, is there any easy way to overcome this? Like settings custom `User-Agent` headers from server side?
I would just go with WebSockets but this limits me from using featurs like **Authentication** and **Identities** etc.
Thank you.

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/dotnet/aspnetcore/issues/30137
Fixes https://github.com/dotnet/aspnetcore/issues/29183

### CODE REVIEW

File: src/SignalR/clients/csharp/Http.Connections.Client/src/HttpConnection.cs [hunk 1]
```diff
@@ -608,6 +607,7 @@ private HttpClient CreateHttpClient()
                     // Check if the key is User-Agent and remove if empty string then replace if it exists.
                     if (string.Equals(header.Key, Constants.UserAgent, StringComparison.OrdinalIgnoreCase))
```
Comment (BrennanConroy) [inline src/SignalR/clients/csharp/Http.Connections.Client/src/HttpConnection.cs:608]: In WASM, the user would need to set "X-SignalR-User-Agent", whereas outside of WASM they need to set "User-Agent" to override the default we set. I could see us keeping this behavior, or saying "User-Agent" will apply to both.

**Comment by davidfowl** [comment]
No test?

**Review by davidfowl — APPROVED** [review]

File: src/Components/test/testassets/BasicTestApp/SignalRClientComponent.razor [hunk 2]
```diff
@@ -47,6 +50,12 @@
         await hubConnection.SendAsync("SendMessage", "SignalR Client", $"Echo {transportType}");
     }
 
+    protected async Task GetUserAgentHeader()
+    {
+        userAgent = await hubConnection.InvokeAsync<string>("GetHeaderValue", "X-SignalR-User-Agent");
```
Comment (davidfowl) [inline src/Components/test/testassets/BasicTestApp/SignalRClientComponent.razor:55]: Seems like you want both user agents right? To make sure the normal is set as well.

Comment (BrennanConroy) [reply src/Components/test/testassets/BasicTestApp/SignalRClientComponent.razor:55]: Normal is set by the browser, we shouldn't be testing browser behavior

Comment (davidfowl) [reply src/Components/test/testassets/BasicTestApp/SignalRClientComponent.razor:55]: So before our code was just ineffective

Comment (BrennanConroy) [reply src/Components/test/testassets/BasicTestApp/SignalRClientComponent.razor:55]: In WASM it was ineffective in some browsers, and set in others
Details at https://github.com/dotnet/aspnetcore/issues/29183#issue-782700972

**Pull request merged**